db/view/view.cc

```diff
@@ -448,44 +452,122 @@ row_marker view_updates::compute_row_marker(const clustering_row& base_row) cons
         default:
             auto& c = update.cells().cell_at(base_col->id);
             auto value_view = base_col->is_atomic() ? c.as_atomic_cell(cdef).value() : c.as_collection_mutation().data;
-            return value_view;
+            return {managed_bytes_view{value_view}};
         }
     });
-    auto& partition = partition_for(partition_key::from_range(_view->partition_key_columns() | get_value));
-    auto ckey = clustering_key::from_range(_view->clustering_key_columns() | get_value);
-    return partition.clustered_row(*_view, std::move(ckey));
+
+    std::vector<view_updates::view_row_entry> ret;
+
+    std::vector<std::vector<bytes_view_with_action>> pk_elems, ck_elems;
+    boost::copy(_view->partition_key_columns() | get_value, std::back_inserter(pk_elems));
+    // If no collection column was found, each of the actions will contain no_action,
+    // in particular, it does not harm to use column 0.
+    size_t action_column = collection_column_position.value_or(0);
+    // Allow for at most one collection computed column in pk and in ck.
+    collection_column_position.reset();
+    boost::copy(_view->clustering_key_columns() | get_value, std::back_inserter(ck_elems));
+
+
+    auto cartesian_product_pk = cartesian_product(pk_elems),
+         cartesian_product_ck = cartesian_product(ck_elems);
+    auto ck_it = cartesian_product_ck.begin();
+
+    auto compute_row = [&](std::vector<bytes_view_with_action>& pk, std::vector<bytes_view_with_action>& ck) {
+        partition_key pkey = partition_key::from_range(boost::adaptors::transform(pk, bytes_view_with_action::get_view));
+        clustering_key ckey = clustering_key::from_range(boost::adaptors::transform(ck, bytes_view_with_action::get_view));
+        auto action = (action_column < pk.size() ? pk[action_column] : ck[action_column - pk.size()])._action;
+
+        mutation_partition& partition = partition_for(std::move(pkey));
+        ret.push_back({&partition.clustered_row(*_view, std::move(ckey)), action});
+    };
+    if (collection_column_position.has_value()) {
+        // The computed collection column in clustering key was associated with the computed collection column from the partion key.
+        for (std::vector<bytes_view_with_action>& pk : cartesian_product_pk) {
+            if (ck_it == cartesian_product_ck.end()) {
+                size_t pk_size = cartesian_product_size(pk_elems),
+                       ck_size = cartesian_product_size(ck_elems);
+                on_internal_error(vlogger, format("Computed sizes of possible partition keys and clustering keys don't match: {} != {}", pk_size, ck_size));
+            }
+            compute_row(pk, *ck_it);
+            ++ck_it;
+        }
+    } else {
+        for (std::vector<bytes_view_with_action>& pk : cartesian_product_pk) {
+            for (std::vector<bytes_view_with_action>& ck : cartesian_product_ck) {
+                compute_row(pk, ck);
```

Comment:
Yes, you're right.

Reply:
I might as well change the type of the internal vector to utils::small_vector, as it will commonly have just one element.

Reply:
To clarify, this condition `if (getter.collection_column_position.has_value()) {` will actually only be true for indexes on values (which are special due to keeping one more clustering key column and handling duplicates later), and it won't be true for collection indexes created on entries/keys, where we have uniqueness? This is definitely not clear from the code itself, so let's add a short comment explaining which branch of this statement has which assumptions. 

I think I get it now, but perhaps it's still worth it to extract a third branch, just for regular indexes, which does not perform any cartesian product computation, because it's guaranteed to have a single element anyway? My line of thought is that collection indexes are generally an edge case, a "slow path" for indexes, and it would be nice to keep the "regular" path less complicated, as it's way more common. But if extracting this third, cartesian-less branch is not trivial and/or makes the code ugly, I'm fine with accepting the code as is too.

Reply:
re `if (getter.collection_column_position.has_value())`: yes, this affects only indexes over collections.

Reply:
I'll extract that third branch shortly.